### PR TITLE
[Merged by Bors] - feat: port Computability.PartrecCode

### PR DESCRIPTION
**Pull request by Komyyy**
This PR also renames some decls in `Computability.Primrec`.

---

What the hell?

- [x] depends on: #3830
- [x] depends on: #3839

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3830
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
> What the hell?

Could you explain what the problem is?

**Comment by semorrison** [comment]
Would you mind pointing to the place where #3839 makes a difference?

**Comment by Komyyy** [comment]
@semorrison This means nothing more than that porting this file is difficult. Thank you.

**Comment by semorrison** [comment]
Okay I was hoping for some explanation that would save me having to open this PR, revert #3839, see for myself what breaks, all before being able to hit merge on #3839. :-)

**Comment by Komyyy** [comment]
The lambda abstraction which is newly inserted in `Nat.rec_zero` breaks `rw` in `Nat.Partrec.Code.eval_prec_zero`.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3830~~
* ~~leanprover-community/mathlib4#3839~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Computability/PartrecCode.lean [hunk 1]
```diff
@@ -292,80 +353,108 @@ theorem rec_prim' {α σ} [Primcodable α] [Primcodable σ] {c : α → Code} (h
...
-              (pc a (of_nat code m.unpair.1, of_nat code m.unpair.2, s₁, s₂)))
-            (cond n.div2.bodd (co a (of_nat code m.unpair.1, of_nat code m.unpair.2, s₁, s₂))
-              (pr a (of_nat code m.unpair.1, of_nat code m.unpair.2, s₁, s₂)))
+            (cond n.div2.bodd (rf a (ofNat Code m, s))
+              (pc a (ofNat Code m.unpair.1, ofNat Code m.unpair.2, s₁, s₂)))
+            (cond n.div2.bodd (co a (ofNat Code m.unpair.1, ofNat Code m.unpair.2, s₁, s₂))
+              (pr a (ofNat Code m.unpair.1, ofNat Code m.unpair.2, s₁, s₂)))
   have : Primrec G₁ := by
-    refine' option_bind (list_nth.comp (snd.comp fst) (snd.comp snd)) _
+    refine' Primrec.option_bind (Primrec.list_get?.comp
```
[the remark this answers is not in the document]
Comment (Komyyy) [reply Mathlib/Computability/PartrecCode.lean:361]: I think that `open Primrec` in L290 actually opens `Nat.Primrec` so many `Primrec`s are required.

[the remark this answers is not in the document]
Comment (jcommelin) [reply Mathlib/Computability/PartrecCode.lean:361]: Asked about this on zulip: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/opening.20namespaces

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**